src/equation_systems/equation_system.C

```diff
@@ -411,25 +410,39 @@ TimeDependentEquationSystem::SetTimeStep(double dt)
 }
 
 void
-TimeDependentEquationSystem::AddKernel(const std::string & test_var_name,
-                                       std::shared_ptr<MFEMBilinearFormKernel> blf_kernel)
+TimeDependentEquationSystem::AddKernel(std::shared_ptr<MFEMKernel> kernel)
 {
-  if (blf_kernel->getTrialVariableName() == GetTimeDerivativeName(test_var_name))
+  if (kernel->getTrialVariableName() == GetTimeDerivativeName(kernel->getTestVariableName()))
   {
+    auto trial_var_name = kernel->getTrialVariableName();
+    auto test_var_name = kernel->getTestVariableName();
     AddTestVariableNameIfMissing(test_var_name);
     AddTrialVariableNameIfMissing(test_var_name);
-    addKernelToMap<MFEMBilinearFormKernel>(blf_kernel, _td_blf_kernels_map);
+    if (!_td_kernels_map.Has(test_var_name))
+    {
+      auto kernel_field_map =
+          std::make_shared<platypus::NamedFieldsMap<std::vector<std::shared_ptr<MFEMKernel>>>>();
+      _td_kernels_map.Register(test_var_name, std::move(kernel_field_map));
+    }
+    // Register new kernels map if not present for the test/trial variable
+    // pair
+    if (!_td_kernels_map.Get(test_var_name)->Has(test_var_name))
+    {
+      auto kernels = std::make_shared<std::vector<std::shared_ptr<MFEMKernel>>>();
+      _td_kernels_map.Get(test_var_name)->Register(test_var_name, std::move(kernels));
+    }
+    _td_kernels_map.GetRef(test_var_name).Get(test_var_name)->push_back(std::move(kernel));
```

Comment:
The comment says you are looking for at the test/trial variable pair, but you only actually work with the case where test and trial variable are the same.